Compile gocean1.0 tests

hiker opened the pull request:
This is for #281.

Note that I don't have easy access to an opencl compiler, so the opencl compilation part is not tested.

I also did not properly address the fact that various combinations of compilers and options might be required for a thorough compilations test (e.g. to test openmp, openacc, opencl) code. I will open a separate issue to find out how we want to support this best.

This is a refactor of the compile environment, wrapping compilation for various APIs into separate classes. The infrastructure libraries will be compiled (once per process if compiling in parallel to avoid race conditions). Oh, I might have sneaked a sentence or two about coverage reporting into the docs while I was updating it for the compilation handling.

arporter commented:
Hi @hiker - is this one ready for review?

hiker commented:
Oops - I was certain I had set the 'ready for review' flag ... only I didn't ;)  Yes, it is ready for a first round :)

arporter commented on src/psyclone/tests/dynamo0p3_build.py:
```diff
@@ -32,34 +32,101 @@
+                                              fort_file)
+                # Abort if there should be a problem
+                if not self.compile_file(name):
+                    break
```
`Dynamo0p3Build._infrastructure_built` will still get set to `True` if this break statement is executed. Should we raise an Exception of some sort instead?

arporter commented on src/psyclone/tests/gocean1p0_build.py:
```diff
@@ -0,0 +1,139 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2017-2018, Science and Technology Facilities Council
```
Possibly this should just be 2019 as this is a new file?

arporter commented on src/psyclone/tests/gocean1p0_build.py:
```diff
@@ -0,0 +1,139 @@
+
+
+''' Module containing configuration required to build code generated
+for the Dynamo0p3 API '''
```
Gocean1p0!

arporter commented on src/psyclone/tests/gocean1p0_build.py:
```diff
@@ -0,0 +1,139 @@
+
+class GOcean1p0Build(Compile):
+    '''Build class for compiling test files for the GOcean1.0 api. It
+    uses dl_esm_inf as included in the PSyclone distribution.
```
"... distribution (as a git submodule)."

arporter commented on src/psyclone/tests/gocean1p0_build.py:
```diff
@@ -0,0 +1,139 @@
+    uses dl_esm_inf as included in the PSyclone distribution.
+    '''
+
+    _infrastructure_built = False
```
Could these be in the base class?

hiker replied:
I prefer them to be there, since it will make it easier (for #314) to handle the case where only one of the APIs will be compiled. Also, if that variable would be shared, I can't see how both infrastructure libraries/wrapper) will be compiled.

arporter replied:
Oh yes, good point.

arporter commented on src/psyclone/tests/gocean1p0_build.py:
```diff
@@ -0,0 +1,139 @@
+
+        # On first instantiation (triggered by conftest.infra_compile)
+        # compile the infrastructure library files.
+        if Compile.TEST_COMPILE and not GOcean1p0Build._infrastructure_built:
```
What about TEST_OPENCL_COMPILE?

arporter commented on src/psyclone/tests/gocean1p0_build.py:
```diff
@@ -0,0 +1,139 @@
+        # compile the infrastructure library files.
+        if Compile.TEST_COMPILE and not GOcean1p0Build._infrastructure_built:
+            self._build_infrastructure()
+            GOcean1p0Build._infrastructure_path = str(tmpdir)
```
It would make sense for `_build_infrastructure` to be responsible for storing the infrastructure path.

arporter commented on src/psyclone/tests/gocean1p0_build.py:
```diff
@@ -0,0 +1,139 @@
+
+    def _build_infrastructure(self):
+        '''Compile dl_esm_inf.
+        '''
```
Please document the return values and the exceptions that can be raised.

arporter commented on src/psyclone/tests/gocean1p0_build.py:
```diff
@@ -0,0 +1,139 @@
+        arg_list = ["make", "F90={0}".format(self._f90),
+                    "F90FLAGS={0}".format(self._f90flags),
+                    "-f", "{0}/Makefile".format(dl_esm_inf_path)]
+        try:
```
The remainder of this method looks very similar to the equivalent in the Dynamo0p3 version of this class. Could the running of the compilation itself and the following error checking be moved into the base class? However, if that would be complicated then don't worry about it.

hiker replied:
The code in Gocean1p0Build._build_infrastructure shares some commonality with Compile._compile_file, not with Dynamo0p3Build - hmm - well, maybe a little bit, catching errors, but I don't see how this could easily be done in one base class (gocean calls popen with build, dynamo calls compile_file in a loop). Maybe a ``print_error`` function or so depending on the exception, but this would likely apply to a lot of other code as well, way beyond the scope of this pull request. So tbh, I'd say for now to just leave it the way it is.

arporter replied:
OK.

arporter commented on src/psyclone/tests/gocean1p0_build.py:
```diff
@@ -0,0 +1,139 @@
+        # Check the return code
+        stat = build.returncode
+        if stat != 0:
+            print(output)
```
Write to stderr?

arporter commented on src/psyclone/tests/gocean1p0_build.py:
```diff
@@ -0,0 +1,139 @@
+        produced are deleted.
+
+        :param psy_ast: The AST of the generated PSy layer
+        :type psy_ast: Instance of :py:class:`psyGen.PSy`
```
s/psyGen.PSy/psyclone.psyGen.PSy/

arporter commented on src/psyclone/tests/gocean1p0_build.py:
```diff
@@ -0,0 +1,139 @@
+            return True
+
+        # Call the internal _code_compile function, since it does
+        # not
```
"since it does not" ....what? :-)

arporter commented on src/psyclone/tests/gocean1p0_opencl_test.py:
```diff
@@ -176,3 +207,4 @@ def test_set_arg_const_scalar():
         otrans.apply(sched)
     assert ("Cannot generate OpenCL for Invokes that contain kernels with "
             "arguments passed by value" in str(err))
+    assert GOcean1p0OpenCLBuild(tmpdir).code_compiles(psy)
```
You don't need this line as the point of this test is that we refuse to generation OpenCL code.

arporter commented on src/psyclone/tests/gocean1p0_transformations_test.py:
```diff
@@ -1242,10 +1266,28 @@ def test_module_inline_same_kernel():
+    # No compilation test here, see test_module _inline_and_compile
+
+
+@pytest.mark.xfail(reason="Inline function uses a module variable")
```
Please could you create an associated Issue for this - e.g. "Validate method of kernel-inline transformation does not check for module variables" - and reference it here.

hiker replied:
Done, #315

arporter replied:
Thanks. Please could you put the ref to #315 inside the `reason` string of the xfail.

arporter commented on src/psyclone/tests/gocean1p0_transformations_test.py:
```diff
@@ -1257,7 +1299,7 @@ def test_module_inline_warning_no_change():
 
 
-def test_loop_swap_correct():
+def ctest_loop_swap_correct(tmpdir):
```
s/ctest/test/ otherwise this test won't be picked up by pytest.

arporter commented on src/psyclone/tests/gocean1p0_transformations_test.py:
```diff
@@ -1405,6 +1447,8 @@ def test_ocl_apply():
 
     gen = str(psy.gen)
     assert "USE clfortran" in gen
+    # ATM no support for opencl compilation
```
Is that true? Actually, I guess it is because OpenCL support in dl_esm_inf is not yet on master? If so, please could you create an issue for this and reference it here with a TODO.

hiker replied:
Now, that's a bit of an issue then: I've just added support for opencl compilation, and assumed the failures were caused by me not having an opencl compiler ... while it is apparently because of #316. Should I now go and remove --compileopencl support etc?? That would be quite a bit of work, which we would have to add back in again when doing #314. I've added a reference to #316 in this one case, but I am not sure what to do with then gocean1p0_opencl_test.py - should we comment out all those tests then?

arporter replied:
I'm happy to leave that flag in because, as you say, it's a lot of work and we will need it in the future.

arporter commented on src/psyclone/tests/test_files/gocean1p0/kernel_stencil.f90:
```diff
@@ -67,11 +67,13 @@ module kernel_stencil
 
   !> Compute the mass flux in the x direction at point (i,j)
-  subroutine compute_cu_code(i, j, cu, p, u)
+  subroutine compute_cu_code(i, j, cu, p, u, area)
```
This kernel doesn't access the `area` array?

hiker replied:
No, it doesn't - but the kernel declaration contained an additional argument, and the test case needs the fourth parameter. I've added a comment explaining why area is there as a paramter.

arporter replied:
Oh yes; the meta-data says it needs the cell area. Don't know why I didn't see that last time around.

hiker commented:
...
> gocean1p0_transformations_test.py:1738: AssertionError
> --------------------------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------------------------
> b"/home/me/Projects/GungHo/PSyclone/src/psyclone/tests/test_files/gocean1p0/kernel_scalar_float.f90:71:8:\n\n use model_mod, only: rdt\n 1\nFatal Error: Can't open module file \xe2\x80\x98model_mod.mod\xe2\x80\x99 for reading at (1): No such file or directory\ncompilation terminated.\n"

Now that one is actually a complete pain. OK, the bug itself is easy to fix, there is a 'use model_mod, only: rdt' statement in kernel_scalar_float, and 'rdt' is unused. So we can easily remove the use statement and all is good.

And it compiled for me, because (from previous work) I had some left-over .mod files in the gocean Fortran test_files directory, and 'model_mod' was one of them. 

After deleting the .mod file I got the same failed test... good :)

But then I checked where else `model_mod` is used, and found `kernel_scalar_int.f90`, which is used by `test_module_inline_with_sub_use `(parsing `single_invoke_scalar_int_arg.f90`). Surprisingly, this test still worked ... because it didn't try to compile in the first place. So I added a compile test here (which then of course failed, since it still could not find model_mod.mod), removed the 'use' statement, and replaced the variable 'used' with a simple constant. While this should compile now, the actual test fails before compilation, since it asserts that this 'use' statement is still in the produced code. And I'd say that this is an important test to make sure we are not losing some use statements somewhere.

So to properly solve this problem, I started to add code to search the kernels for other use statements, which could then be added to the list of modules that need to be compiled (see `kernel_modules` that must be compiled) before the psy-layer can be. But then the problem of course is that a kernel also 'uses' dl_esm_inf modules, which it then tries to compile (it can't easily know if a module is coming from dl_esm_inf, or the application). 

As you can see, things are getting rather complicated then (we would need to only compile 'use'd modules that are not part of dl_esm_inf, so more tests there, ...).

My suggestion: we just do not compile files that must have a 'use' statement in them (like kernel_scalar_int.f90). In many cases where the 'use' is not essential for the test I will just remove the 'use' statement and replace any variable with a simple constant or so.

Is that ok? I'd say our compilation tests will be quite thorough anyway, if there is a handful(?) of Fortran files that are not tested for compilation, I'd think that's acceptable.

> 
> I'm also struggling to remember why we thought that OpenCL compilation support needed to be separate. The FortCL library is just Fortran. OpenCL only becomes an issue at link time. However, once @sergisiso has OpenCL kernel generation working we will need to be able to verify the compilation of those. Perhaps that's what we were thinking of?

The OpenCL Fortran files have a 'use fortcl ..' statement, which we can't compile. Since I would suspect that many users have a setup similar to mine (gfortran, no opencl), I wanted to make sure that this setup works.

The rest is work in progress :)

arporter commented:
> My suggestion: we just do not compile files that must have a 'use' statement in them (like kernel_scalar_int.f90). In many cases where the 'use' is not essential for the test I will just remove the 'use' statement and replace any variable with a simple constant or so.
> 
> Is that ok? I'd say our compilation tests will be quite thorough anyway, if there is a handful(?) of Fortran files that are not tested for compilation, I'd think that's acceptable.
> 
Yes, that sounds fine to me. We need to tackle the whole issue of use statements in kernels anyway so it's OK to punt it for the moment.

> 
> The OpenCL Fortran files have a 'use fortcl ..' statement, which we can't compile. Since I would suspect that many users have a setup similar to mine (gfortran, no opencl), I wanted to make sure that this setup works.
> 
Ah, I see. FortCL is a submodule of dl_esm_inf but again, that's probably not yet on master of dl_esm_inf. Once it is I think compilation should be straightforward, at least for the Fortran driver stuff. The transformed kernels will be a different matter.

hiker commented on doc/developers.rst:
```diff
@@ -90,14 +90,26 @@ you would do::
+  > py.test --cov-report term-missing --cov-report html --cov psyclone.dynamo0p3 ./dynamo0p3_basis_test.py ./parse_test.py
+
+The html output can be viewed with a browser at ``file:///.../tests/htmlcov/index.html``
+and it highlights all source lines in red that are not covered by at least one test.
```
OK, if the used library for opencl will be part of dl_esm_inf, we can fix the compiles. But on the other hand, we need to handle opencl separate anyway, so I'd recommend to leave the separate command line option in place, and handle the rest later (e.g. as part of #314),

hiker commented on doc/developers.rst:
```diff
@@ -147,10 +166,30 @@ specified by further command-line flags::
+process will compile its own version of the files to avoid race conditions.
+This happens only once per process in each test session.
+
+For the gocean1.0 API a complete copy of the dl_esm_inf library is included in
```
Oops - fixed :)

hiker commented on doc/profiling.rst:
```diff
@@ -220,9 +220,16 @@ As an example::
+library. Support for further profiling libraries will be
+added in the future. To compile the wrapper libraries,
+change into the directory ``lib/profiling`` of PSyclone
+and type ``make`` to compile all wrapper. If only some
```
Yes, I fixed the typo, but that paragraph will be changed in #297.

hiker commented on src/psyclone/tests/conftest.py:
```diff
@@ -86,6 +91,28 @@ def have_graphviz():
     return "graphviz" in sys.modules
 
 
+@pytest.fixture(scope="session", autouse=True)
```
*grin* I spent a lot of time reading about fixtures :)

hiker commented on src/psyclone/tests/psyclone_test_utils.py:
```diff
@@ -107,206 +98,261 @@ def print_diffs(expected, actual):
+        if not link:
+            arg_list.append("-c")
+        if self.get_infrastructure_flags():
+            arg_list += self.get_infrastructure_flags()
```
Hmm - yes and no ;) In my original implementation get_infrastructure_flags was abstract, but later I found that utilities_test best uses Compile directly, which was why I needed the if test. I now changed Compile.get_infrastructure_flags() to return [], then I could remove the if test.

hiker commented on src/psyclone/tests/psyclone_test_utils.py:
```diff
@@ -107,206 +98,261 @@ def print_diffs(expected, actual):
+        :rtype: bool
+
+        '''
+        if not Compile.TEST_COMPILE:
```
Good question :) I think they all end up using compile_file, which tested both, but it is of course much better to test earlier.

hiker commented on src/psyclone/tests/psyclone_test_utils.py:
```diff
@@ -107,206 +98,261 @@ def print_diffs(expected, actual):
+    def compile_file(self, filename, link=False):
+        ''' Compiles the specified Fortran file into an object file (in
+        the current working directory) using the specified Fortran compiler
+        and flags. Raises a CompileError if the compilation fails.
```
It turns out that compile_file ALWAYS returns true (or raises an exception). So I've removed the return value completely here and in all tests.

arporter replied:
Great.

hiker commented:
I got stuck on a train today due to signal failure ... 1.5 hours later I have some unplanned, minor additional improvements, I am just not sure if you want them included here or in a separate request. The first part is strongly related to compilation, and the second part is strongly related to the first part, but not to compilation 😄   Not sure if this is a good enough argument though to include them both here - so you decide.

1) I removed the directory cleanup after compilation. *VERY* often I had to look at the actual files to see what the reason for a test failure is, or compare with a previous run - and always got stuck because the cleanup had removed those files. So I ended up constantly commenting out the cleanup. Note that py.test manages the temporary directories for you, it deletes all but the last three ones when you start it. And I admit I don't understand the comment in the sources that the directory cleanup will allow the same directory to be used several times - even if the files in there you can compile several times(??) - what am I missing? Anyway, I have a patch that removes the cleanup (and also removes the test cases that check if the directories are cleaned up).
2) I replaced all use of tempfiles in tests with tmpdir fixtures (and then also removed the deletion of those files). In case you haven't noticed, there were at least some tempfile tests that leaked files in /tmp ... I had 260 files there 😏  Now everything is under the py.test temporary directory, and automatically cleaned up.

arporter commented:
Hi @hiker. I thought that kind of signal failure only happened here - it's kind of reassuring that it's not just a British thing :-) Those changes you describe sound like they fit OK with this PR. TBH, I can't remember quite what I was worrying about when I said that a test could safely re-use a directory. Possibly I was (mistakenly) thinking of a make-type system that would not bother compiling a file if the .o already existed. Then again, maybe it was something else. Either way I'm happy with your proposed solution as it uses pytest functionality and leaves the generated files for debugging purposes.
Is this PR ready for review again or was that just a comment checking whether or not to include these new changes in it?

hiker commented:
A British thing?? No way, I am sure the Australians invented signal failures!

OK, I've added the temporary file handling improvements. And I need the updated dl_esm_inf (https://github.com/stfc/dl_esm_inf/pull/24) so that all opencl tests will actually work.
*grin* While submodule are really nice, it makes for a lot of work to get an update in: I needed an update to FortCL (to compile in separate directory), which then required an update to dl_esm_inf (to compile fortcl and itself in separate directory), so that I can then finally update psyclone to compile the dl_esm_inf per process. I feel like Marco Polo on my way East ... it's a long trip ;)

arporter commented:
Thinking about it, maybe we introduced signal failures into Australia :-) It is, as you say, a long trip. I think that's just because we have a lot of dependencies and many of them are in a state of flux. If you can think of any way of improving things then all suggestions are welcome :-)

hiker commented:
OK, ready for a review. Note that I had several problems with the update, likely because of the nested submodules we now have (could also be because I am using a windows file system mounted on a linux VM - issues with permissions). If a 'git status' aborts with some kind of porcelain errors, try a mixture of `git submodule init`, and `git submodule update --init --recursive`. At a certain stage it worked for me.

arporter commented on src/psyclone/tests/psyclone_test_utils.py:
```diff
@@ -160,27 +152,28 @@ def base_path(self, base_path):
         the module files of the infrastructure can be found.
         :returns: A list of strings with the compiler flags required.
-        :rtpe: str
+        :rtpe: list
```
s/rtpe/rtype/

arporter commented on src/psyclone/tests/psyclone_test_utils.py:
```diff
@@ -107,206 +98,261 @@ def print_diffs(expected, actual):
+
+    @property
+    def base_path(self):
+        '''Returns the base path of all test files for the API. Needs to
```
Sorry to be a pain but it's still not clear to me what this means. Is it the location of the Fortran infrastructure files that need to be compiled?

arporter replied:
Actually, from a sub-class I think it's the location of the compiled infrastructure files? Please put this in the docstring.

arporter commented:
Oh yes, I too hit that problem with git - thanks for the tip. Please could you add it to the developers' guide in the relevant section?

arporter commented on src/psyclone/tests/psyclone_test_utils.py:
```diff
@@ -340,17 +326,12 @@ def string_compiles(self, code):
+        success = True
         try:
-            success = self.compile_file(filename)
+            self.compile_file(filename)
```
Now that you've simplified this, you could simplify it further by just doing `return False` in the `except CompileError` part and `return True` at the end of the routine.

arporter replied:
(and pylint complains that old_pwd is not used)

hiker replied:
Actually ... the warning correctly indicates that I am not changing back to the original directory. While I don't think this will break anything (afaik any code using a temporary directory will chdir first anyway), but as long as we don't have this as a requirement, I think it's better that I add a finally and chdir back to tmpdir - which then also means keeping the success variable around.

arporter replied:
OK, I'm happy with that - it seems the safest thing to do.

arporter commented on src/psyclone/tests/psyclone_test_utils.py:
```diff
@@ -272,29 +265,22 @@ def _code_compiles(self, psy_ast):
+            self.compile_file(psy_filename)
         except CompileError:
             # Failed to compile one of the files
             success = False
```
Just `return False` and then `return True` below?

arporter replied:
Also, pylint now complains that old_pwd is not used.

arporter commented on src/psyclone/tests/dynamo0p3_build.py:
```diff
@@ -42,9 +42,11 @@
-    '''Build class for compiling test files for the Dyanmo0.3 api.
-    using the wrapper library contained in the test_files.
+    '''Build class for compilation of test files for the Dynamo0.3 api.
+    Is uses the wrapper library form test_files/dynamo0p3/infrastructure
```
s/Is uses the wrapper library form/It uses the wrapper library from/

hiker commented:
OK, ready again ... I hope :)

arporter commented:
Thanks for making those changes and updating the documentation.
All comments have been addressed.
Documentation builds OK and looks OK.
No xfailing tests are affected.
Coverage of the diff is NA in this case (only the testing framework is affected).
